server/frontend.go

```diff
@@ -3,68 +3,99 @@ package server
 import (
 	"log"
 	"os"
+	"sync/atomic"
+	"time"
 
 	"github.com/privacylab/talek/common"
 )
 
 // Frontend terminates client connections to the leader server.
+// It is the point of global serialization, and establishes sequence numbers.
 type Frontend struct {
 	// Private State
-	log          *log.Logger
-	name         string
-	serverConfig *Config
-	follower     common.FollowerInterface
-	isLeader     bool
+	log  *log.Logger
+	name string
+	*Config
 
-	//dataLayerRef *DataLayerRef
+	proposedSeqNo uint64 // Use atomic.AddUint64, atomic.LoadUint64
+	readChan      chan *readRequest
+
+	replicas []common.ReplicaInterface
+	dead     int
+}
+
+// readRequest is the grouped request and reply memory used for batching
+// incoming reads onto a single thread.
+type readRequest struct {
+	Args  *common.EncodedReadArgs
+	Reply *common.ReadReply
+	Done  chan bool
 }
 
 // NewFrontend creates a new Frontend for a provided configuration.
-func NewFrontend(name string, serverConfig *Config, follower common.FollowerInterface, isLeader bool) *Frontend {
+func NewFrontend(name string, config *Config, replicas []common.ReplicaInterface) *Frontend {
 	fe := &Frontend{}
 	fe.log = log.New(os.Stdout, "[Frontend:"+name+"] ", log.Ldate|log.Ltime|log.Lshortfile)
 	fe.name = name
-	fe.serverConfig = serverConfig
-	fe.follower = follower
-	fe.isLeader = isLeader
+	fe.Config = config
+	fe.replicas = replicas
+	fe.readChan = make(chan *readRequest, 10)
+
+	// Periodically serialize database epoch advances.
+	go fe.periodicWrite()
+	// Batch incoming reads into combined requests to replicas.
+	go fe.batchReads()
 
 	return fe
 }
 
 /** PUBLIC METHODS (threadsafe) **/
 
-// Ping implements latency testing determination.
-func (fe *Frontend) Ping(args *common.PingArgs, reply *common.PingReply) error {
-	fe.log.Println("Ping: " + args.Msg + ", ... Pong")
-
-	// Try to ping the follower if one exists
-	if fe.follower != nil {
-		var fReply common.PingReply
-		fErr := fe.follower.Ping(&common.PingArgs{Msg: "PING"}, &fReply)
-		if fErr != nil {
-			var fName string
-			fe.follower.GetName(nil, &fName)
-			reply.Err = fName + " Ping failed"
-		} else {
-			reply.Err = fReply.Err
-		}
-	} else {
-		reply.Err = ""
-	}
+// Close goroutines associated with this object.
+func (fe *Frontend) Close() {
```

Comment:
Need better concurrency control here if we have any state modified by multiple goroutines. 
e.g.
func (fe *Frontend) Close() {
	atomic.StoreInt32(&fe.dead, 1)
}
func (fe *Frontend) isDead() bool {
	return atomic.LoadInt32(&fe.dead) != 0
}